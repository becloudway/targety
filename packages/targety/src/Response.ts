/* tslint:disable:object-literal-sort-keys */
import {
    ApiError,
    BadRequestError,
    ConflictError,
    ForbiddenError,
    InternalServerError,
    NotFoundError,
    UnauthorizedError,
    ValidationError,
} from "./errors";
import { Strings } from "./utils";
import cookie from "cookie";
import { ContentType } from "./common/enums/ContentType";
import { Request } from "./Request";
import { ResponseBody, ResponseBodyInput } from "./ResponseBody";
import yn from "yn";
import { URL } from "url";

const { API_VERSION, ALLOW_LOCALHOST } = process.env;

export interface ResponseOptions {
    cors?: boolean;
    defaultHeaders?: boolean;
}

export interface CookieOptions {
    httpOnly?: boolean;
    secure?: boolean;
    domain?: string;
    path?: string;
    sameSite?: "lax" | "strict" | "none";
    expires?: Date;
}

interface JsonObject {
    [key: string]: any;
}

/**
 * Wrapper for ApiGateway Response
 *
 * ```
 * // { statusCode: 200, body: { success: true }};
 * return Response.ok(request, { success: true });
 *
 * // { statusCode: 403, body: { error: ... }};
 * const forbiddenError = new ForbiddenError();
 * return Response.fromError(request, forbiddenError);
 *
 * // { statusCode: 302, body: '' } with redirect header;
 * const forbiddenError = new ForbiddenError();
 * return Response.redirect(request, 'https://redirect/to');
 * ```
 */
export class Response<T> {
    private allowedOrigins: string[] = Strings.toList(process.env.ALLOWED_ORIGINS || "*");
    private body: T; // ResponseBodyInput['body'] = '';
    private status: ResponseBodyInput["statusCode"];
    private headers: ResponseBodyInput["headers"];
    private contentType: string;
    private cors = true;
    private defaultHeaders = true;
    private baseUrl: string;
    private allowLocalhost: boolean = yn(ALLOW_LOCALHOST);

    constructor(statusCode: number, request: Request, options?: ResponseOptions) {
        this.status = statusCode;
        this.cors = options && options.cors === false ? options.cors : this.cors;
        this.defaultHeaders = options?.defaultHeaders === false ? options.defaultHeaders : this.defaultHeaders;
        if (this.cors) {
            this.setCorsHeaders(request);
        }

        if (this.defaultHeaders) {
            this.setDefaultHeaders(request);
        }
    }

    public static render(request: Request, options?: ResponseOptions): Response<string> {
        // don't load the default headers on a render as they might interfere with the rendered page
        const response = Response.response<string>(200, request, { defaultHeaders: false, ...options });
        response.setHeader("Content-Type", ContentType.HTML);
        return response;
    }

    public static redirect(request: Request, redirectUrl: string, options?: ResponseOptions): Response<null> {
        // don't load the default headers on a redirect as they might interfere with the frontend
        const response = Response.response<null>(302, request, { defaultHeaders: false, ...options });
        response.setHeader("location", redirectUrl);
        response.setHeader("content-type", ContentType.HTML);
        return response;
    }

    public static ok(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(200, request, options);
    }

    public static accepted(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(202, request, options);
    }

    public static created(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(201, request, options);
    }

    public static noContent(request: Request, options?: ResponseOptions): Response<null> {
        return Response.response<null>(204, request, options);
    }

    public static badRequest(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(400, request, options);
    }

    public static conflict(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(409, request, options);
    }

    public static notFound(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(404, request, options);
    }

    public static internalServerError(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(500, request, options);
    }

    public static unauthorized(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(401, request, options);
    }

    public static forbidden(request: Request, options?: ResponseOptions): Response<JsonObject> {
        return Response.response<JsonObject>(403, request, options);
    }

    public static fromError(request: Request, error: ApiError | Error, options?: ResponseOptions): ResponseBody {
        const convertError = (err: ApiError | Error, defaultMessage?: string) => {
            const errorMessage = err.message || defaultMessage;
            return errorMessage
                ? {
                      message: errorMessage,
                      errorCode: (err as ApiError).errorCode,
                      awsRequestId: request.getRequestId(),
                  }
                : err;
        };

        if (error instanceof InternalServerError) {
            return Response.internalServerError(request, options).send(convertError(error));
        }

        if (error instanceof ConflictError) {
            return Response.conflict(request, options).send(convertError(error));
        }

        if (error instanceof NotFoundError) {
            return Response.notFound(request, options).send(convertError(error));
        }

        if (error instanceof UnauthorizedError) {
            return Response.unauthorized(request, options).send(convertError(error));
        }

        if (error instanceof BadRequestError) {
            return Response.badRequest(request, options).send(convertError(error));
        }

        if (error instanceof ForbiddenError) {
            return Response.forbidden(request, options).send(convertError(error, "Forbidden"));
        }

        if (error instanceof ValidationError) {
            return Response.badRequest(request, options).send({
                message: "Validation error",
                errorCode: "ValidationError",
                ...JSON.parse(error.message),
                awsRequestId: request.getRequestId(),
            });
        }

        return Response.internalServerError(request, options).send(convertError(error));
    }

    private static corsResponse(request: Request, allowedOrigins: string[]): string {
        const origin = request.getOrigin();
        if (!origin) {
            throw new InternalServerError("Origin request header not set");
        }
        if (allowedOrigins.includes("*") || allowedOrigins.length === 0 || allowedOrigins.includes(origin)) {
            return origin;
        }
        return null;
    }

    private static response<T>(statusCode: number, request: Request, options?: ResponseOptions): Response<T> {
        return new Response<T>(statusCode, request, options);
    }

    /**
     * Sets contentType based on body and stringifies the body before returning [[ResponseBody]]
     */
    public send(body: T | string = null): ResponseBody {
        body = body == null ? this.body : body;

        if (!this.getHeader("content-type")) {
            this.setHeader("content-type", "text");
            this.contentType = "text";
        }

        if (typeof body === "object") {
            this.setHeader("content-type", "application/json");
            return this.send(JSON.stringify(body));
        }

        if (typeof body !== "string") {
            body = (JSON.stringify(body) as unknown) as string;
        }

        const responseBody = new ResponseBody({
            body: (body as unknown) as string,
            headers: this.headers,
            statusCode: this.status,
        });
        return responseBody;
    }

    public addCookie(key: string, val: string, options: CookieOptions = {}): Response<T> {
        const cookieHeader = this.headers["set-cookie"] || [];
        const cookieOptions: CookieOptions = {
            httpOnly: true,
            secure: true,
            domain: new URL(this.baseUrl).host,
            path: "/",
            sameSite: "none",
            ...options,
        };
        cookieHeader.push(cookie.serialize(key, val, cookieOptions));
        this.headers["set-cookie"] = cookieHeader;

        return this;
    }

    /**
     * Set header `field` to `val`, or pass
     * an object of header fields.
     *
     * Examples:
     *
     *    res.set('Foo', ['bar', 'baz']);
     *    res.set('Accept', 'application/json');
     *    res.set({ Accept: 'text/plain', 'X-API-Key': 'tobi' });
     *
     * Aliased as `res.header()`.
     *
     * @param {String|Object} field
     * @param {String|Array} [val]
     * @return {Response} for chaining
     * @public
     */
    public setHeader(field: string, val: string | string[]): Response<T> {
        const value: string[] = val ? (Array.isArray(val) ? val.map((v) => `${v}`) : [`${val}`]) : [];

        if (typeof this.headers === "undefined") {
            this.headers = {};
        }

        this.headers[field.toLowerCase()] = value;

        return this;
    }

    /**
     * Removes a header from the headers object
     *
     * @param field the field to be removed
     */
    public removeHeader(field: string): void {
        if (this.headers[field]) {
            delete this.headers[field];
        }
    }

    /**
     * Get value for header `field`.
     *
     * @param {String} field
     * @return {String}
     * @public
     */
    private getHeader(field: string): string | undefined {
        return typeof this.headers !== "undefined"
            ? this.headers[field.toLowerCase()] && this.headers[field.toLowerCase()][0]
            : undefined;
    }

    /**
     * Sets some default headers, that can be overridden if needed.
     * Loosely based on: https://cheatsheetseries.owasp.org/cheatsheets/REST_Security_Cheat_Sheet.html
     */
    private setDefaultHeaders(request: Request): void {
        if (!this.allowLocalhost) {
            this.setHeader("Strict-Transport-Security", "max-age=31536000");
        }

        this.setHeader("X-Frame-Options", "DENY");
        this.setHeader("Cache-Control", "no-store");
        this.setHeader("X-Content-Type-Options", "nosniff");
        this.setHeader("Content-Security-Policy", "'self'");

        this.setHeader("X-AWS-Request-Id", request.getRequestId());
    }

    private setCorsHeaders(request: Request): void {
        this.setHeader("Access-Control-Allow-Origin", Response.corsResponse(request, this.allowedOrigins));
        this.setHeader("Access-Control-Allow-Credentials", "true");
    }

    public setHeaders(headers: ResponseBodyInput["headers"]): void {
        for (const header in headers) {
            if (Object.prototype.hasOwnProperty.call(headers, header)) {
                this.setHeader(header, headers[header]);
            }
        }
    }

    public setBaseUrl(baseUrl: string): void {
        this.baseUrl = baseUrl;
    }
}
