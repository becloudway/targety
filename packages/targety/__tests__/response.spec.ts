/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/ban-ts-ignore */
import { ApiGateWayProxyEvent } from "@dev/test-helper";
import {
    BadRequestError,
    ConflictError,
    ForbiddenError,
    InternalServerError,
    NotFoundError,
    UnauthorizedError,
    ValidationError,
} from "../src/errors";
import { ContentType } from "../src/common/enums/ContentType";
import { LambdaProxyEvent, Request } from "../src/Request";
import { Response } from "../src/Response";

let request: Request;
describe("Response", () => {
    beforeEach(() => {
        request = new Request((ApiGateWayProxyEvent.get() as unknown) as LambdaProxyEvent);
    });

    describe("cookies", () => {
        it("adds a cookie", () => {
            const response = Response.redirect(request, "http://example.com/callback");
            response.setBaseUrl("http://example.com/api");
            response.addCookie("accessToken", "1234", { secure: true, httpOnly: true });
            const responseBody = response.send();
            expect(responseBody.multiValueHeaders["set-cookie"]).toHaveLength(1);
            expect(responseBody.multiValueHeaders["set-cookie"][0]).toMatch("accessToken=1234");
            expect(responseBody.multiValueHeaders["set-cookie"][0]).toMatch("Secure");
            expect(responseBody.multiValueHeaders["set-cookie"][0]).toMatch("HttpOnly");
        });
        it("allows multiple cookies", () => {
            const response = Response.redirect(request, "http://example.com/callback");
            response.setBaseUrl("http://example.com/api");
            response.addCookie("accessToken", "1234", { secure: true, httpOnly: true });
            response.addCookie("refreshToken", "1234", { secure: true, httpOnly: true, path: "/auth/refresh" });
            const responseBody = response.send();
            expect(responseBody.multiValueHeaders["set-cookie"]).toHaveLength(2);
        });
    });
    describe("CORS", () => {
        describe("With cors disabled allow requests without origin", () => {
            it("cors disabled: allows absence of origin header", () => {
                const response = Response.noContent(request, { cors: false }).send();
                expect(response.statusCode).toEqual(204);
            });
            it("cors enabled (default): does not allow absence of origin header", () => {
                // @ts-ignore
                request.getOriginClean = jest.fn().mockReturnValue(undefined);
                request.getOrigin = jest.fn().mockReturnValue(undefined);
                expect(() => Response.noContent(request).send()).toThrow(InternalServerError);
            });
        });
        describe("Allowed headers are not specified", () => {
            it('returns cors header "access-control-allow-origin" with request origin', () => {
                // @ts-ignore
                request.headers.origin = ["http://localhost:8000"];
                const response = Response.ok(request).send({ success: "ok" });
                expect(response.multiValueHeaders["access-control-allow-origin"]).toContain("http://localhost:8000");
            });
        });
        describe("Allowed headers are specified", () => {
            it('returns cors header "access-control-allow-origin" with request origin', () => {
                // @ts-ignore
                process.env.ALLOWED_ORIGINS = "http://localhost:8000";
                const { Response: ResponseWithCors } = require("../src/Response");
                // @ts-ignore
                request.headers.origin = ["http://foreign/domain"];
                const response = ResponseWithCors.ok(request).send({ success: "ok" });
                expect(response.multiValueHeaders["access-control-allow-origin"]).toEqual([]);
            });
            it('returns cors header "access-control-allow-origin" with origin domain when mulitpel cors headers are allowed', () => {
                // @ts-ignore
                process.env.ALLOWED_ORIGINS = "http://localhost:8000,http://allowed/domain";
                // @ts-ignore
                const { Response: ResponseWithCors } = require("../src/Response");
                // @ts-ignore
                request.headers.origin = ["http://allowed/domain"];
                const response = ResponseWithCors.ok(request).send({ success: "ok" });
                expect(response.multiValueHeaders["access-control-allow-origin"]).toContain("http://allowed/domain");
            });
        });
    });

    describe("Should set default headers", () => {
        it("should set all the default headers", () => {
            const response = Response.ok(request).send({ success: "ok" });
            expect(response).toMatchSnapshot();
        });
    });

    describe("2XX", () => {
        it("#ok: sends 200 response with json", () => {
            const response = Response.ok(request).send({ success: "ok" });
            expect(response.statusCode).toEqual(200);
            expect(response.multiValueHeaders["content-type"]).toContain(ContentType.JSON);
            expect(JSON.parse(response.body as string)).toEqual({ success: "ok" });
        });
        it("#render: sends an 200 response with html", () => {
            const responseBody = "<h1>header</h1>";
            const response = Response.render(request).send(responseBody);
            expect(response.statusCode).toEqual(200);
            expect(response.multiValueHeaders["content-type"]).toContain(ContentType.HTML);
            expect(response.body).toEqual(responseBody);
        });
        it("#created: sends 201 response with json", () => {
            const response = Response.created(request).send({ success: "ok" });
            expect(response.statusCode).toEqual(201);
            expect(response.multiValueHeaders["content-type"]).toContain(ContentType.JSON);
            expect(JSON.parse(response.body as string)).toEqual({ success: "ok" });
        });
        it("#noContent: sends 204 response with empty body", () => {
            const response = Response.noContent(request).send();
            expect(response.statusCode).toEqual(204);
            expect(response.body).toEqual("");
        });
    });
    describe("3XX", () => {
        it("#redirect: sends 302 with location header", () => {
            const response = Response.redirect(request, "redirectUrl").send();
            expect(response.statusCode).toEqual(302);
            expect(response.multiValueHeaders.location).toContain("redirectUrl");
        });
    });

    describe("4XX", () => {
        it("#badRequest: sends a 400 with error body", () => {
            const expectedStatusCode = 400;
            const apiError = new BadRequestError("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
        it("#validationError: sends a 400 with error body", () => {
            const expectedStatusCode = 400;
            const apiError = new ValidationError({
                validationErrors: [
                    {
                        field: "name",
                        validationError: "name is required",
                    },
                ],
            });
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual("Validation error");
            expect(errorBody.validationErrors).toEqual(JSON.parse(apiError.message).validationErrors);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
        it("#unauthorized: sends a 401 with error body", () => {
            const expectedStatusCode = 401;
            const apiError = new UnauthorizedError("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
            expect(errorBody.stack).toBeUndefined();
        });
        it("#forbidden: sends a 403 with error body", () => {
            const expectedStatusCode = 403;
            const apiError = new ForbiddenError("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
        it("#notFound: sends a 404 with error body", () => {
            const expectedStatusCode = 404;
            const apiError = new NotFoundError("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
        it("#conflict: sends a 409 with error body", () => {
            const expectedStatusCode = 409;
            const apiError = new ConflictError("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
    });
    describe("5XX", () => {
        it("#internalServer: sends a 500 with error body", () => {
            const expectedStatusCode = 500;
            const apiError = new InternalServerError("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toEqual(apiError.errorCode);
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
        it("#internalServer: sends a 500 with error body when error is not an apiError", () => {
            const expectedStatusCode = 500;

            const apiError = new Error("Oops");
            const response = Response.fromError(request, apiError);
            const errorBody = JSON.parse(response.body as string);

            expect(response.statusCode).toEqual(expectedStatusCode);
            expect(errorBody.message).toEqual(apiError.message);
            expect(errorBody.errorCode).toBeUndefined();
            expect(errorBody.awsRequestId).toEqual(request.getRequestId());
        });
    });
});
