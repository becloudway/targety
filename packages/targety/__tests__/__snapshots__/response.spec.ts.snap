// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response Should set default headers should set all the default headers 1`] = `
ResponseBody {
  "body": "{\\"success\\":\\"ok\\"}",
  "multiValueHeaders": Object {
    "access-control-allow-credentials": Array [
      "true",
    ],
    "access-control-allow-origin": Array [],
    "cache-control": Array [
      "no-store",
    ],
    "content-security-policy": Array [
      "'self'",
    ],
    "content-type": Array [
      "application/json",
    ],
    "strict-transport-security": Array [
      "max-age=31536000",
    ],
    "x-aws-request-id": Array [
      "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
    ],
    "x-content-type-options": Array [
      "nosniff",
    ],
    "x-frame-options": Array [
      "DENY",
    ],
  },
  "statusCode": 200,
}
`;
